util/util.go

```diff
@@ -50,6 +51,31 @@ func StringInSlice(s string, slice []string) bool {
 	return util.StringInSlice(s, slice)
 }
 
+func StringToImageReference(image string) (types.ImageReference, error) {
```

Comment:
`UIArgumentToImageReference`? `HeuristicUserInputToImageReference`? Something along those lines? (Up to Buildah maintainers.)

I, fairly desperately, don’t want this code to be used to “round-trip” between `ImageReference` and `string` [there already are other functions for that], and this plain naming of the function suggests that it could be used for that purpose.

Reply:
> Some sort of abuse this function would get by being used in unintended places?

Yes. The function is a heuristic that can parse the input using several different format/semantics, depending on what happens to work.

A failure case I’ve seen fairly frequently in this space is finding some pre-existing function like that, similarly targeted at the UI and heuristically doing “what the user meant”, and embedding it somewhere deeper in the call stack, calling it with a string produced from a Go value, _without accounting for the details of the heuristics_, or possibly the CLI-level heuristics being expanded to add new user-level features. The outcome is that the Go value → string → Go value transformation (which might, sometimes, be unavoidable, e.g. for multi-machine operations) becomes inexact, hard to predict, and possibly even leads to security issues.

Hence all of the concern about _not using_ strings as an internal representation if at all possible, as the easiest preventative measure.

Secondarily, using the right string representation / API. The `types.ImageReference` values already have an API for producing a string representation, and for creating one from a string representation. That one (`alltransports.ParseImageName`) should be used anywhere below the CLI heuristic layers, and the new function should be named so that it isn’t accidentally used instead.

So that’s why I would like some mention of UI, and preferably “heuristics” in the name.

“Image string” does not help because it does not disambiguate “a lossy heuristic subject to change” from “a stable round-trip-safe value conversion”, either would apply to images and strings.